File: std/string.lua
-- DEAL Standard Library: std/string — v1.2
-- Provides string manipulation functions per the v1.2 spec.
-- String lengths and positions are measured in Unicode scalar values.

local __rt = require("deal.runtime")

local stringlib = {}

--- Escape Lua pattern magic characters so the string can be used
-- as a literal in string.gsub (which lacks a plain-text flag).
local function escape_pattern(s)
  return (s:gsub("([%^%$%(%)%%%.%[%]%*%+%-%?])", "%%%1"))
end

--- Escape Lua gsub replacement magic characters (%0–%9, %%).
-- gsub interprets % sequences in the replacement string,
-- so a caller-supplied "to" parameter must be escaped to
-- prevent unintended substitutions.
local function escape_replacement(s)
  return (s:gsub("%%", "%%%%"))
end

--- length(s: string): int
-- Returns the number of Unicode scalar values in the string (v1.2).
stringlib.length = __rt.function_("(string)->int", function(s)
  __rt.check_string(s)
  local count = 0
  local cursor = 0
  while true do
    local next_cursor = __rt.utf8_next(s, cursor)
    if next_cursor == nil then break end
    count = count + 1
    cursor = next_cursor
  end
  return __rt.check_int(count)
end)

--- substring(s: string, start: int, end: int): string
-- Returns the substring from start (0-based, inclusive) to end (exclusive),
-- with positions measured in Unicode scalar values (v1.2).
-- start > end, start beyond the string, or end <= 0 produce the empty
-- string; out-of-range end clamps to the string end.
stringlib.substring = __rt.function_("(string,int,int)->string", function(s, start, end_)
  __rt.check_string(s)
  __rt.check_int(start)
  __rt.check_int(end_)
  -- Walk scalar values to translate scalar positions to byte offsets.
  local cursor = 0
  local pos = 0
  while pos < start do
    local next_cursor = __rt.utf8_next(s, cursor)
    if next_cursor == nil then break end
    cursor = next_cursor
    pos = pos + 1
  end
  local byte_start = cursor + 1
  while pos < end_ do
    local next_cursor = __rt.utf8_next(s, cursor)
    if next_cursor == nil then break end
    cursor = next_cursor
    pos = pos + 1
  end
  local byte_end = cursor
  return __rt.check_string(string.sub(s, byte_start, byte_end))
end)

--- contains(s: string, part: string): boolean
-- Returns true if part is a substring of s (plain-text match, no pattern).
stringlib.contains = __rt.function_("(string,string)->boolean", function(s, part)
  __rt.check_string(s)
  __rt.check_string(part)
  -- Use plain-text matching (4th arg to string.find)
  local found = string.find(s, part, 1, true)
  return __rt.check_boolean(found ~= nil)
end)

--- startsWith(s: string, part: string): boolean
-- Returns true if s starts with part.
stringlib.startsWith = __rt.function_("(string,string)->boolean", function(s, part)
  __rt.check_string(s)
  __rt.check_string(part)
  -- Empty part is a prefix of any string
  if #part == 0 then
    return __rt.check_boolean(true)
  end
  return __rt.check_boolean(#s >= #part and s:sub(1, #part) == part)
end)

--- endsWith(s: string, part: string): boolean
-- Returns true if s ends with part.
stringlib.endsWith = __rt.function_("(string,string)->boolean", function(s, part)
  __rt.check_string(s)
  __rt.check_string(part)
  -- Empty part is a suffix of any string
  if #part == 0 then
    return __rt.check_boolean(true)
  end
  -- s:sub(-0) would return the whole string; guard with the length check
  if #part > #s then
    return __rt.check_boolean(false)
  end
  return __rt.check_boolean(s:sub(-#part) == part)
end)

--- replace(s: string, old: string, to: string): string
-- Replaces all occurrences of old with to (plain-text, no pattern).
-- Uses string.gsub with escaped pattern and escaped replacement for
-- plain-text semantics.
stringlib.replace = __rt.function_("(string,string,string)->string", function(s, old, to)
  __rt.check_string(s)
  __rt.check_string(old)
  __rt.check_string(to)
  if old == "" then
    return __rt.check_string(s)
  end
  local escaped_pattern = escape_pattern(old)
  local escaped_to = escape_replacement(to)
  return __rt.check_string((s:gsub(escaped_pattern, escaped_to)))
end)

--- split(s: string, sep: string): [string]
-- Splits s by sep (plain-text) and returns an array of substrings.
-- If sep is empty, splits into individual characters.
-- If sep is not found, returns an array containing s as the single element.
-- The wrapper signature carries the canonical descriptor grammar
-- (luajit-v1.2-stdlib-contracts D1): arrays are "[T]".
stringlib.split = __rt.function_("(string,string)->[string]", function(s, sep)
  __rt.check_string(s)
  __rt.check_string(sep)
  local result = {}

  -- Empty string: return empty array regardless of separator
  if #s == 0 then
    return result
  end

  if sep == "" then
    -- Split into individual Unicode scalar values (v1.2).
    local cursor = 0
    while true do
      local next_cursor, ch = __rt.utf8_next(s, cursor)
      if next_cursor == nil then break end
      result[#result + 1] = ch
      cursor = next_cursor
    end
    return result
  end

  local start_pos = 1
  while true do
    local found_start, found_end = string.find(s, sep, start_pos, true) -- plain-text
    if not found_start then
      result[#result + 1] = s:sub(start_pos)
      break
    end
    result[#result + 1] = s:sub(start_pos, found_start - 1)
    start_pos = found_end + 1
  end

  return result
end)

--- trim(s: string): string
-- Removes leading and trailing whitespace from s.
-- Whitespace includes space, tab, newline, carriage return, vertical tab, form feed.
stringlib.trim = __rt.function_("(string)->string", function(s)
  __rt.check_string(s)
  -- Use Lua pattern: ^%s* captures leading whitespace, (.-) lazily captures content, %s*$ captures trailing
  local trimmed = string.match(s, "^%s*(.-)%s*$")
  if trimmed == nil then
    -- string.match returns nil when there is no match; but ^%s*(.-)%s*$ always matches
    -- (even on empty string it matches with empty capture). Fallback for safety.
    return __rt.check_string("")
  end
  return __rt.check_string(trimmed)
end)

return stringlib

File: std/string.js
"use strict";

// DEAL Standard Library: std/string — v1.2 (JavaScript backend).
// The hand-written CommonJS string module (js-stdlib-modules D1/D5,
// js-backend-runtime D9): the eight .d.deal members exported as
// $rt.function wrappers with the exact spec signature strings, parameter
// checks inside each body, and the declared-return check at exit — every
// check receiving the forwarded trailing span parameters ($file, $line,
// $column), so parameter and return errors report the DEAL call site.
// String lengths and positions are measured in Unicode scalar values
// (code points, v1.2); search/replace/split are plain-text (never Lua
// patterns); trim strips the Lua %s set (space, tab, newline, vertical
// tab, form feed, carriage return) exactly.
//
// Host-global capture: none (js-stdlib-modules D3 — the module uses
// String prototype methods and $rt scalar helpers only; no bare global
// is spelled). The runtime is required with the fixed ../deal/runtime
// specifier, which resolves <output>/deal/runtime.js from the deployed
// <output>/std/ location and the repo-root deal/runtime.js from the
// repo-root std/ directory (the dual-location loadability of
// js-stdlib-modules D1). The module is pure and stateless: inputs are
// never mutated, no shared state, no I/O.

const $rt = require("../deal/runtime");

// $isTrimChar: the Lua %s character set — space, tab, newline, vertical
// tab, form feed, carriage return (the six C-locale isspace characters
// Lua 5.1 pattern %s matches). Never String.prototype.trim, whose set is
// wider (js-stdlib-modules D5). All six characters are single-unit
// ASCII, so a UTF-16-unit test is exact and never cuts a multi-unit
// scalar.
function $isTrimChar($c) {
  return $c === 0x20 || $c === 0x09 || $c === 0x0a || $c === 0x0b || $c === 0x0c || $c === 0x0d;
}

const $string = {};

// length(s: string): int — the number of Unicode scalar values in s
// (the code-point count; scalarLength walks one code point per step).
$string.length = $rt.function("(string)->int", function(s, $file, $line, $column) {
  $rt.checkString(s, $file, $line, $column);
  return $rt.checkInt($rt.scalarLength(s), $file, $line, $column);
});

// substring(s: string, start: int, end: int): string — the substring at
// 0-based scalar positions [start, end) with the v1.2 clamping
// (js-stdlib-modules D5): negative start behaves as 0; negative end
// behaves as 0 (yields ""); end beyond the scalar length clamps to the
// length; start >= end yields ""; start beyond the string yields "".
// Observably identical to the reference's two scalar-position walks
// feeding string.sub (std/string.lua substring), whose i > j /
// negative-j / beyond-length corrections produce exactly these
// outcomes. The collection is one code-point walk over [start, end)
// after the clamping.
$string.substring = $rt.function("(string,int,int)->string", function(s, start, end, $file, $line, $column) {
  $rt.checkString(s, $file, $line, $column);
  $rt.checkInt(start, $file, $line, $column);
  $rt.checkInt(end, $file, $line, $column);
  const $len = $rt.scalarLength(s);
  let $st = start < 0 ? 0 : start;
  let $en = end < 0 ? 0 : end;
  if ($en > $len) $en = $len;
  if ($st >= $en) {
    return $rt.checkString("", $file, $line, $column);
  }
  let $result = "";
  let $pos = 0;
  for (const $c of s) {
    if ($pos >= $en) break;
    if ($pos >= $st) $result += $c;
    $pos++;
  }
  return $rt.checkString($result, $file, $line, $column);
});

// contains(s: string, part: string): boolean — plain-text substring
// presence via s.includes(part): a string argument is literal, never a
// pattern; the empty part is contained in every string (the reference's
// plain-text string.find). Equality is exact because equal scalar
// sequences have equal UTF-16 encodings (js-backend-runtime D8) — no
// helper needed.
$string.contains = $rt.function("(string,string)->boolean", function(s, part, $file, $line, $column) {
  $rt.checkString(s, $file, $line, $column);
  $rt.checkString(part, $file, $line, $column);
  return $rt.checkBoolean(s.includes(part), $file, $line, $column);
});

// startsWith(s: string, part: string): boolean — plain-text prefix test
// via s.startsWith(part); the empty part is a prefix of every string.
$string.startsWith = $rt.function("(string,string)->boolean", function(s, part, $file, $line, $column) {
  $rt.checkString(s, $file, $line, $column);
  $rt.checkString(part, $file, $line, $column);
  return $rt.checkBoolean(s.startsWith(part), $file, $line, $column);
});

// endsWith(s: string, part: string): boolean — plain-text suffix test
// via s.endsWith(part); the empty part is a suffix of every string and a
// longer part yields false.
$string.endsWith = $rt.function("(string,string)->boolean", function(s, part, $file, $line, $column) {
  $rt.checkString(s, $file, $line, $column);
  $rt.checkString(part, $file, $line, $column);
  return $rt.checkBoolean(s.endsWith(part), $file, $line, $column);
});

// replace(s: string, old: string, to: string): string — plain-text
// replacement of every occurrence via s.split(old).join(to): a string
// separator to split is literal, never a pattern, and a % in to is
// literal (no gsub capture references). The old === "" guard returns s
// unchanged — the reference's explicit return-s-unchanged branch
// (std/string.lua replace).
$string.replace = $rt.function("(string,string,string)->string", function(s, old, to, $file, $line, $column) {
  $rt.checkString(s, $file, $line, $column);
  $rt.checkString(old, $file, $line, $column);
  $rt.checkString(to, $file, $line, $column);
  if (old === "") {
    return $rt.checkString(s, $file, $line, $column);
  }
  return $rt.checkString(s.split(old).join(to), $file, $line, $column);
});

// split(s: string, sep: string): string[] — plain-text split
// (js-stdlib-modules D5): empty s → [] whatever the separator (the
// reference's #s == 0 first branch); empty sep → $rt.scalars(s), one
// scalar per part (a supplementary character is one element); otherwise
// s.split(sep) — not-found → [s], and leading/trailing/consecutive
// separators produce empty elements exactly like the reference's
// plain-text string.find loop (a trailing separator yields the empty
// last element). The exit check validates the array and each string
// element.
$string.split = $rt.function("(string,string)->[string]", function(s, sep, $file, $line, $column) {
  $rt.checkString(s, $file, $line, $column);
  $rt.checkString(sep, $file, $line, $column);
  if (s === "") {
    return $rt.checkArray("[string]", [], $file, $line, $column);
  }
  if (sep === "") {
    return $rt.checkArray("[string]", $rt.scalars(s), $file, $line, $column);
  }
  return $rt.checkArray("[string]", s.split(sep), $file, $line, $column);
});

// trim(s: string): string — strips the leading and trailing Lua %s set
// (the six characters of $isTrimChar) via two UTF-16-unit index walks,
// then slices. Internal whitespace is preserved; an all-whitespace
// input and the empty input yield "".
$string.trim = $rt.function("(string)->string", function(s, $file, $line, $column) {
  $rt.checkString(s, $file, $line, $column);
  let $a = 0;
  const $n = s.length;
  while ($a < $n && $isTrimChar(s.charCodeAt($a))) {
    $a++;
  }
  let $b = $n;
  while ($b > $a && $isTrimChar(s.charCodeAt($b - 1))) {
    $b--;
  }
  return $rt.checkString(s.slice($a, $b), $file, $line, $column);
});

module.exports = $string;
